__tests__/implementation.test.js

```diff
@@ -292,6 +292,117 @@ test('batch endpoint (multiple requests)', async () => {
+                expect(include_extra_info).toBe(true);
+                return Promise.resolve([
+                    {
+                        foo_id: 3,
+                        rating: 5,
+                        name: 'Shake Shack',
+                        extra_stuff: 'lorem ipsum',
+                    },
+                ]);
+            }
+        },
+    };
+
+    await createDataLoaders(config, async (getLoaders) => {
+        const loaders = getLoaders(resources);
+
+        const results = await loaders.foo.loadMany([
+            { foo_id: 2, property: 'name', include_extra_info: false },
+            { foo_id: 1, property: 'rating', include_extra_info: false },
+            { foo_id: 3, property: 'rating', include_extra_info: true },
+        ]);
+
+        expect(results).toEqual([
+            { foo_id: 2, rating: 4, name: 'In N Out' },
+            { foo_id: 1, rating: 3, name: 'Burger King' },
+            { foo_id: 3, rating: 5, name: 'Shake Shack', extra_stuff: 'lorem ipsum' },
+        ]);
+    });
+});
+
+test('batch endpoint (multiple requests) with secondaryBatchKey different response structure', async () => {
+    const config = {
+        resources: {
+            foo: {
+                isBatchResource: true,
+                docsLink: 'example.com/docs/bar',
+                batchKey: 'foo_ids',
+                newKey: 'foo_id',
+                secondaryBatchKey: 'properties',
+                secondaryNewKey: 'property',
+            },
+        },
+    };
+
+    const resources = {
+        foo: ({ foo_ids, properties, include_extra_info }) => {
+            if (_.isEqual(foo_ids, [2, 1])) {
+                expect(include_extra_info).toBe(false);
+                return Promise.resolve([
+                    { 1: { rating: 3, name: 'Burger King' } },
+                    { 2: { rating: 4, name: 'In N Out' } },
+                ]);
+            }
+
+            if (_.isEqual(foo_ids, [3])) {
+                expect(include_extra_info).toBe(true);
+                return Promise.resolve([
+                    {
+                        3: {
+                            rating: 5,
+                            name: 'Shake Shack',
+                            extra_stuff: 'lorem ipsum',
+                        },
+                    },
+                ]);
+            }
+        },
+    };
+
+    await createDataLoaders(config, async (getLoaders) => {
+        const loaders = getLoaders(resources);
+
+        const results = await loaders.foo.loadMany([
+            { foo_id: 2, property: 'name', include_extra_info: false },
+            { foo_id: 1, property: 'rating', include_extra_info: false },
+            { foo_id: 3, property: 'rating', include_extra_info: true },
+        ]);
+
+        expect(results).toEqual([
+            { 2: { rating: 4, name: 'In N Out' } },
+            { 1: { rating: 3, name: 'Burger King' } },
+            { 3: { rating: 5, name: 'Shake Shack', extra_stuff: 'lorem ipsum' } },
```

Comment:
Hmm, for now, we cannot just get what we asked for. 
That seems related to your data masking questions in the previous comment, assuming we have following service call:
```
getBusiness({ foo_ids: [1,2], properties: ['rating', 'name'], include_extra_info: false});
getBusiness({ foo_ids: [3], properties: ['rating'], include_extra_info: true});
````
And we only requested the name property like this `{ foo_id: 2, property: 'name', include_extra_info: false },` , it will return both `rating` and `name` data. 

I think it's fine to return extra stuff, people usually need to do a little transformation of what they get from dataloader anyway. And in this case, when they get data from dataloader in the resolver, they could simply check if the property they asked for is there and then return that property.  
It's the same idea as this resolver, 
```
async name(parent, args, context) {
  const { name } = await context.serviceDataLoaders.getBusinesses.load({ businessId});
  return name;
```
Here, `getBusinesses` endpoint returns not only `name` but lots of other things, like `photos`, `address`, `country`, etc., but the `name` resolver only need the `name` field.

Reply:
> Here, getBusinesses endpoint returns not only name but lots of other things, like photos, address, country, etc., but the name resolver only need the name field.

The interface of this API is well documented. It always returns you that stuff. To change that would be a breaking change, which is how we protect callsites.

The interface of `getBusinessProperties` (for example) is similarly well documented to say "if you ask for property foo, we return you foo.".

What's concerning here is someone could do:

```js
const { foo, bar } = await context.loaders.getBusinessProperties({ property: ['foo'] });
```

...and this would work (note that we didn't ask for 'bar', but we can get it back because someone else asked for it!)

This seems like a footgun for two reasons:
- This only works in a situation where a different resolver called `getBusinessProperties({ property: ['bar'] });` previously. A dev may test using a particular query that sets up this behaviour - but when the resolver gets called in a different query without the resolver that called bar, then boom - now this logic breaks.
- More generally, we could have devs writing code assuming a different API than what the endpoint really does. The dataloader layer is "lying" to resolver authors, and no longer fully transparent. If ever we decide to try and fix this behavior, too bad! resolvers are already relying on this broken logic.

tl;dr this is brittle and potentially dangerous

Reply:
Fair enough, I will try to see if we could only return the property we ask for.

Reply:
@magicmark I was able to only return the property we ask for, by removing any other data that is neither `batchKey` nor `secondayBatchKey`, see https://github.com/Yelp/dataloader-codegen/pull/261/files#diff-92b6f52e2c1b302cd9d15133831c595f151215aa1dfda274ff29fbd33d89a9b2R356-R358. 

But that only works if the endpoint doesn't return extra_data for any properties. 
For example, 
`getBusiness({ foo_ids: [1], properties: ['name', 'rating']}) => {foo_id: 1, name: "Burger King", rating: 3}`
and we only requested the name property like this `{ foo_id: 1, property: 'name',}`, it works fine, we will get `{foo_id: 1, name: "Burger King"}`

But if the endpoint not only returns the property data, but also returns some extra data for some property, those extra data will be removed.
For example, 
`getBusiness({ foo_ids: [3], properties: ['rating', 'name']); => {foo_id: 3, rating: 4, name: "In N Out", extra_data: 'blabla', extra_info: 'blabla'}` 
and we only requested the name property like this `{ foo_id: 3, property: 'name'}`, we will get `{foo_id: 3, name: "In N Out"}`. Is this the correct data? I don't know. We just cannot tell the `extra_data` and `extra_info` fields are associated with `rating` or `name` property. From the original use case https://github.com/Yelp/dataloader-codegen/issues/75 and my use case, I don't see the endpoint could return any extra data other than `properties`, so it might be okay to ignore this corner case. What do you think?

Reply:
> But if the endpoint not only returns the property data, but also returns some extra data for some property, those extra data will be removed.

I see - to clarify, this is a problem uniquely when the "properties" are returned embedded in the top level object, rather than in a nested "properties" object - which would be easier to do this logic for?

Without doing something nasty like inspecting the swagger type, I agree the data masking isn't easily possible.

I'm wondering if this is a good reason to switch gears entirely.

Right now, it's impossible to get back more data from internal api than you asked for.

This change would introduce corner cases and for the reasons stated in https://github.com/Yelp/dataloader-codegen/pull/261#discussion_r631284573, i'm not sure we should introduce this behaviour.

I see two options:
1) Only do this logic for endpoints where the "properties" are nested in the response, so we can do the masking logic without removing top level stuff. This would require the endpoints themselves to change.
2) Discard this and do something else for the data hydration endpoints that needed this

How feasible is Option 1? We can tell folks "if you want to batch on a secondary key, the response must be nested".

I would then also rename `secondaryKey` -> `propertyMergeKey` since this is getting pretty behaviour specific, and I don't think we'll ever want tertiary batch key...

Reply:
> Only do this logic for endpoints where the "properties" are nested in the response, so we can do the masking logic without removing top level stuff. 
@magicmark Just to clarify that "properties" are nested in the response` looks like this:
```
{
 'id_1': {
   'name': 'In N Out',
   'rating': 3,
 },
 'extra_stuff': {...},
}
```
NOT this, 
```
{
 'id': 'id_1',
 'name': 'In N Out',
 'rating': 3,
 'extra_stuff': {...},
}
```
I guess we could ask `"if you want to batch on a secondary "poperties" key, then you have to conform to this response shape"` but could also add something like, "otherwise, make sure no extra data is returned."

Reply:
@magicmark I just realized something I didn't think of before and it's harder to deal with
For example, we have the loader get called with the following arguments,
```
loaders.sbp.getBusiness.load({ business_id: 1, properties: ['rating']});
loaders.sbp.getBusiness.load({ business_id: 2, properties: ['rating']});
loaders.sbp.getBusiness.load({ business_id: 3, properties: ['name']});
loaders.sbp.getBusiness.load({ business_id: 1, properties: ['name']});
```
If everything goes well, `business_id` is inside the return object, we could easily match response to request. However, if it throws some errors, we don't know which one to match to. 
```
getBusiness({ business_ids: [1, 2, 3], properties: ['rating', 'name']});
=> [
    CaughtResourceError(.....), 
    { id: 2, rating: 3.5, name: ‘In N Out’ },
    BatchItemNotFoundError(.....),
]
```

Reply:
> Even if the "properties" are nested in the response, it could look like this

I'm suggesting that we _don't_ allow responses that we don't know how to merge. We would have to set up a "property merging contract" that endpoints must implement if they want to be merged by dataloader-codegen.

For this network request:
```
getBusiness({ encids: [1,2], properties: ['rating', 'name'] });
```

The response should look like this:

```js
[
    {
        encid: 1,
        properties: {
            rating: 3.14,
            name: "foo",
        },
    },
    {
        encid: 2,
        properties: {
            rating: 4.99,
            name: "bar",
        },
    },
];
```

This means we can do the data masking of the response purely in the `.properties` field - only allow through the thing that the callsite requested

Reply:
@magicmark This makes sense to me, how about the error situation I mentioned above?

Reply:
>  how about the error situation I mentioned above?

As discussed, we can probs pair on this

Separate thought: let's make sure the folks who end the endpoints we're trying to solve internally are cool with this contract? :)

Reply:
I have submitted a commit that handle the error situation. So no need to pair on this for now.

Reply:
> Separate thought: let's make sure the folks who own the endpoints we're trying to solve internally are cool with this contract? :)

Sure, I will ping them tomorrow.